qi-doc/scribblings/forms.scrbl

```diff
@@ -14,10 +14,152 @@
+
+The symbol @racket[expr] is typically used in this sense to indicate a Racket nonterminal position in the syntax -- that is, a position that expects a Racket expression. Analogously, we use the identifier @deftech{@racket[floe]} (pronounced "flow-e," for "flow expression") to refer to the Qi nonterminal, i.e. a position expecting Qi syntax.
+
+The full syntax of Qi ("Standard Qi") is given below. Note that Standard Qi expands to a @seclink["The_Qi_Core_Language"]{smaller core language} before being @seclink["It_s_Languages_All_the_Way_Down"]{compiled to Racket}.
+
+@racketgrammar*[
+[floe _
+      (gen expr ...)
+      △
+      sep
+      ▽
+      collect
+      (esc expr)
+      (clos floe)
+      (as identifier ...)
+      (one-of? expr ...)
+      (all floe)
+      (any floe)
+      (none floe)
+      (and floe ...)
+      (or floe ...)
+      (not floe)
+      (and% floe ...)
+      (or% floe ...)
+      NOT
+      !
+      AND
+      &
+      OR
+      ∥
+      NOR
+      NAND
+      XOR
+      XNOR
+      any?
+      all?
+      none?
+      inverter
+      ⏚
+      ground
+      (~> floe ...)
+      (thread floe ...)
+      (~>> floe ...)
+      (thread-right floe ...)
+      X
+      crossover
+      ==
+      (== floe ...)
+      relay
+      (relay floe ...)
+      (==* floe ...)
+      (relay* floe ...)
+      -<
+      (-< floe ...)
+      tee
+      (tee floe ...)
+      fanout
+      (fanout nat)
+      feedback
+      (feedback nat floe)
+      (feedback nat (then floe) floe)
+      (feedback (while floe) floe)
+      (feedback (while floe) (then floe) floe)
+      count
+      1>
+      2>
+      3>
+      4>
+      5>
+      6>
+      7>
+      8>
+      9>
+      (select index ...)
+      (block index ...)
+      (bundle (index ...) floe floe)
+      group
+      (group nat floe floe)
+      sieve
+      (sieve floe floe floe)
+      (partition [floe floe] ...)
+      (if floe floe)
+      (if floe floe floe)
+      (when floe floe)
+      (unless floe floe)
+      switch
+      (switch switch-expr ...)
+      (switch (% floe) switch-expr ...)
+      (switch (divert floe) switch-expr ...)
+      (gate floe)
+      ><
+      (>< floe)
+      amp
+      (amp floe)
+      pass
+      (pass floe)
+      <<
+      (<< floe)
+      (<< floe floe)
+      >>
+      (>> floe)
+      (>> floe floe)
+      (loop floe)
+      (loop floe floe)
+      (loop floe floe floe)
+      (loop floe floe floe floe)
+      (loop2 floe floe floe)
+      (ε floe floe)
+      (effect floe floe)
+      apply
+      (qi:* expr ...)
+      (expr expr ... __ expr ...)
+      (expr expr ... _ expr ...)
+      (expr expr ...)
+      literal
+      identifier]
+[literal boolean
```

Comment:
I was surprised to find out that all of these are implemented manually! The knot is tied by the `val:literal` clause in the main branch's `qi-lib/flow/extended/expander.rkt`, which gets a `literal` class from `qi-lib/flow/aux-syntax.rkt`.

Reply:
Do you mean that it's surprising that `literal` isn't provided built-in by Racket? If so, do you think the class as we've defined it would be worth contributing to Racket?

Reply:
> Do you mean that it's surprising that `literal` isn't provided built-in by Racket?

Yeah: I expected there to be an easy way to match literals (maybe by expanding and matching `#%datum`, but I'm not sure I'd call that easy or performant).

> If so, do you think the class as we've defined it would be worth contributing to Racket?

Maybe!

Reply:
Why would matching `#%datum` not be performant?

It also occurred to me that as Qi expansion happens before Racket expansion, we may not be able to use `#%datum` in any case. But I wonder if it would be an option for Racket macros (so that Racket may not need `literal`)? I tried this quick test and it doesn't work though:

```
(require syntax/parse/define)

(define-syntax-parser blah
  [(_ (#%datum v)) #''hi])

((blah 5)) ;=> blah: bad syntax  in: (blah 5)
```

Maybe the issue is that `#%datum` is added at the final stages of Racket macro expansion, so that Racket macros can't use it for matching, either.

Reply:
I wrote

> maybe by expanding and matching #%datum, but I'm not sure I'd call that easy or performant

IOW: you'd have to locally expand the input and _then_ match for `#%datum`, which I expect to be inefficient.

Reply:
Ah sorry I misread that, and yes, that makes sense!